doc/design/tensor_design.md

```diff
@@ -0,0 +1,175 @@
+# Tensor Design
+
+
+## Allocation
+
+`Allocation` is a [RAII](http://en.cppreference.com/w/cpp/language/raii) class template, which is used to handle a piece of memory. 
+
+```cpp
+// Template parametr 'Device' can be 'CpuDevice' or 'GpuDevice'
+template <typename Device>
+class Allocation {
+public:
+    Allocation();
+    Allocation(size_t size, Device device);
+
+    // Creates a non-owned allocation
+    Allocation(void* ptr, size_t size, Device device);
+
+    ~Allocation();
+    //No copying!
+    Allocation(const Allocation&) = delete;
+    //No assigning!
+    Allocation& operator=(const Allocation&) = delete;
+
+    void* ptr() const;
+    void* end() const;
+    Device device() const;
+    size_t size() const;
+
+private:
+    bool owned_;
+    void* ptr_;
+    size_t size_;
+    Device device_;
+};
+```
+
+`ptr_` points to the head of the memory piece and `size_` shows its length. `owned_` marks whether the memory piece is allocated by `allocation` itself, if so the memory will be freed when `allocation` is destructed.
+
+`Device` is something like Majel's `Place`. However, `Place` in Majel is an alias of `boost::variant`, while `Device` here is a certain class (can be specialized to `CpuDevice` or `GpuDevice`). `CpuDevice` and `GpuDevice` are exactly Majel's `CpuPlace` and `GpuPlace`, we rename them to fit the overall naming style.
+
+```cpp
+struct CpuDevice{
+    inline bool operator==(const CpuDevice&) const {
+        return true;
+    }
+    inline bool operator!=(const CpuDevice&) const {
+        return false;
+    }
+};
+
+struct GpuDevice {
+    GpuDevice(int d) : device_id(d) { }
+    
+    inline bool operator==(const GpuDevice &o) const {
+        return device_id == o.device_id;
+    }
+    inline bool operator!=(const GpuDevice &o) const {
+        return !(*this == o);
+    }
+    GpuDevice() : GpuDevice(0) { }
+    int device_id;
+};
+```
+
+## Tensor
+
+`Tensor` is the combination of Majel's `Buffer` and `Array`.
+
+```cpp
+template<typename Device, typename T, int rank>
+class Tensor {
+public:
+    // tensor with zero size and no memory
+    Tensor();
+    // allocates new densely packed tensor
+    Tensor(const Dim<rank> size, Device device);
+
+    // make a new tensor by another existing tensor
+    // new tensor and source tensor have the same numel but deferent rank
+    template<int src_rank>
+    Tensor(const Dim<rank>& size, const Tensor<Device, T, src_rank>& src);
```

Comment:
Is the new tensor should be a writable view of src tensor?

If so, then new tensor can change data which shared with src tensor, i.e., new tensor can change src tensor.  So, src tensor should not be marked as `const`?